crates/ty_python_semantic/src/types/tuple.rs

```diff
@@ -0,0 +1,819 @@
+                Some(generic_context) => Some(
+                    class_literal
+                        .apply_specialization(db, |_| generic_context.specialize_tuple(db, self)),
+                ),
+            })
+    }
+
+    /// Return a normalized version of `self`.
+    ///
+    /// See [`Type::normalized`] for more details.
+    #[must_use]
+    pub(crate) fn normalized(self, db: &'db dyn Db) -> Self {
+        TupleType::new(db, self.tuple(db).normalized(db))
+    }
+
+    pub(crate) fn materialize(self, db: &'db dyn Db, variance: TypeVarVariance) -> Self {
+        TupleType::new(db, self.tuple(db).materialize(db, variance))
+    }
+
+    pub(crate) fn apply_type_mapping<'a>(
+        self,
+        db: &'db dyn Db,
+        type_mapping: &TypeMapping<'a, 'db>,
+    ) -> Self {
+        TupleType::new(db, self.tuple(db).apply_type_mapping(db, type_mapping))
+    }
+
+    pub(crate) fn find_legacy_typevars(
+        self,
+        db: &'db dyn Db,
+        typevars: &mut FxOrderSet<TypeVarInstance<'db>>,
+    ) {
+        self.tuple(db).find_legacy_typevars(db, typevars);
+    }
+
+    pub(crate) fn has_relation_to(
+        self,
+        db: &'db dyn Db,
+        other: Self,
+        relation: TypeRelation,
+    ) -> bool {
+        self.tuple(db)
+            .has_relation_to(db, other.tuple(db), relation)
+    }
+
+    pub(crate) fn is_equivalent_to(self, db: &'db dyn Db, other: Self) -> bool {
+        self.tuple(db).is_equivalent_to(db, other.tuple(db))
+    }
+
+    pub(crate) fn is_gradual_equivalent_to(self, db: &'db dyn Db, other: Self) -> bool {
+        self.tuple(db).is_gradual_equivalent_to(db, other.tuple(db))
+    }
+
+    pub(crate) fn is_disjoint_from(self, db: &'db dyn Db, other: Self) -> bool {
+        self.tuple(db).is_disjoint_from(db, other.tuple(db))
+    }
+
+    pub(crate) fn is_fully_static(self, db: &'db dyn Db) -> bool {
+        self.tuple(db).is_fully_static(db)
+    }
+
+    pub(crate) fn is_single_valued(self, db: &'db dyn Db) -> bool {
+        self.tuple(db).is_single_valued(db)
+    }
+}
+
+/// A fixed-length tuple.
+#[derive(Clone, Debug, Default, Eq, Hash, PartialEq, salsa::Update)]
+pub struct FixedLengthTuple<'db>(Vec<Type<'db>>);
+
+impl<'db> FixedLengthTuple<'db> {
+    pub(crate) fn empty() -> Self {
+        Self::default()
+    }
+
+    pub(crate) fn with_capacity(capacity: usize) -> Self {
+        Self(Vec::with_capacity(capacity))
+    }
+
+    pub(crate) fn from_elements(elements: impl IntoIterator<Item = impl Into<Type<'db>>>) -> Self {
+        Self(elements.into_iter().map(Into::into).collect())
+    }
+
+    pub(crate) fn as_slice(&self) -> &[Type<'db>] {
+        &self.0
+    }
+
+    pub(crate) fn fixed_elements(&self) -> impl Iterator<Item = Type<'db>> + '_ {
+        self.0.iter().copied()
+    }
+
+    pub(crate) fn all_elements(&self) -> impl Iterator<Item = Type<'db>> + '_ {
+        self.0.iter().copied()
+    }
```

Comment:
it feels a bit silly to have two methods that do exactly the same thing -- can't we merge `fixed_elements()` and `all_elements()` into a single `elements()` method? It feels self-evident that they will return the same thing for a `FixedLengthTuple` instance 😆

could we also maybe rename `as_slice()` to `elements_slice()`, to more clearly indicate that it returns a slice of the underlying elements?